### PR TITLE
Option to train just on the data of the last X triggers

### PR DESCRIPTION
**Pull request by francescodeaglio**
By experimenting on the yearbook dataset, we observed that during a distribution shift, recent points are preferred when downsampling techniques are used. With this PR we want to give the possibility to only take the latest triggers into account. 

For this first PR, this parameter is only implemented with NewDataStrategy but the goal is to have it for every strategy

### CODE REVIEW

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py [hunk 1]
```diff
@@ -46,6 +46,20 @@ def __init__(
...
+            ), "Reset after trigger is equivalent to setting tail triggers to 0."
+            assert not (
+                not self.reset_after_trigger and self.tail_triggers == 0
+            ), "Reset after trigger is equivalent to setting tail triggers to 0."
+        else:
+            if self.reset_after_trigger:
+                self.tail_triggers = 0  # consider only the current trigger
+            else:
+                self.tail_triggers = -1  # consider every datapoint
+
```
Comment (MaxiBoether) [inline modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:62]: Please add in all existing strategies checks that self.tail_triggers == -1 or not given at all, so we don't ignore it silently.

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py [hunk 2]
```diff
@@ -46,6 +46,20 @@ def __init__(
...
         self.has_limit = self.training_set_size_limit > 0
         self.reset_after_trigger: bool = config["reset_after_trigger"]
+        if "tail_triggers" in config:
+            self.tail_triggers = config["tail_triggers"]
+            assert not (
+                self.reset_after_trigger and self.tail_triggers > 0
+            ), "Reset after trigger is equivalent to setting tail triggers to 0."
+            assert not (
+                not self.reset_after_trigger and self.tail_triggers == 0
+            ), "Reset after trigger is equivalent to setting tail triggers to 0."
```
Comment (MaxiBoether) [inline modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:56]: Please use ValueErrors for invalid user input. assertions are for invalid internal state

File: modyn/selector/internal/selector_strategies/new_data_strategy.py [hunk 3]
```diff
@@ -86,6 +88,16 @@ def _get_data_reset(self) -> Iterable[list[int]]:
 
             yield samples
 
+    def _get_data_tail(self) -> Iterable[list[int]]:
+        assert not self.reset_after_trigger and self.tail_triggers > 0
+
+        for samples in self._get_tail_triggers_data():
```
Comment (MaxiBoether) [inline modyn/selector/internal/selector_strategies/new_data_strategy.py:94]: Can you please add tests for the NewDataStrategy testing for correct behavior :)? Thanks!

File: modyn/selector/internal/selector_strategies/new_data_strategy.py [hunk 4]
```diff
@@ -131,7 +143,31 @@ def _get_current_trigger_data(self) -> Iterable[list[int]]:
...
+            list[int]: Keys of used samples
+        """
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            stmt = (
+                select(SelectorStateMetadata.sample_key)
+                # Enables batching of results in chunks. See https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-yield-per
+                .execution_options(yield_per=self._maximum_keys_in_memory)
+                .filter(
+                    SelectorStateMetadata.pipeline_id == self._pipeline_id,
+                    self._next_trigger_id - SelectorStateMetadata.seen_in_trigger_id <= self.tail_triggers,
```
Comment (MaxiBoether) [inline modyn/selector/internal/selector_strategies/new_data_strategy.py:170]: Could you rephrase this as conditions on SelectorStateMetadata? Like SelectorStateMetadata. seen_in_trigger_id >= - self.tail_triggers - self._next_trigger_id or something that is maybe a bit more readable? It's a bit more easy to read to condition on SelectorStateMetadata I think.

File: modyn/selector/internal/selector_strategies/new_data_strategy.py [hunk 5]
```diff
@@ -131,7 +143,31 @@ def _get_current_trigger_data(self) -> Iterable[list[int]]:
                 .execution_options(yield_per=self._maximum_keys_in_memory)
                 .filter(
                     SelectorStateMetadata.pipeline_id == self._pipeline_id,
-                    SelectorStateMetadata.seen_in_trigger_id == self._next_trigger_id,
+                    self._next_trigger_id == SelectorStateMetadata.seen_in_trigger_id,
```
Comment (MaxiBoether) [inline modyn/selector/internal/selector_strategies/new_data_strategy.py:146]: Could you reverse the swap here?

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thanks for the PR! I have some suggestions, see below :) Thank you!!

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py (hunk 1 above)
Comment (francescodeaglio) [reply modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:63]: I implemented it on all methods except freshness sampling, on which I put the check

File: modyn/selector/internal/selector_strategies/new_data_strategy.py (hunk 5 above)
Comment (francescodeaglio) [reply modyn/selector/internal/selector_strategies/new_data_strategy.py]: Sure!

**Review by MaxiBoether — APPROVED** [review]
I have one minor suggestion of not using -1 as a magic number - everything else looks good!
